dll/win32/shell32/brfolder.cpp

```diff
@@ -804,6 +805,84 @@ BrFolder_OnCommand(BrFolder *info, UINT id)
     }
 }
 
+static void
+GetTreeViewItemContextMenuPos(HWND hWnd, HTREEITEM hItem, POINT *ppt)
+{
+    RECT rc;
+    if (TreeView_GetItemRect(hWnd, hItem, &rc, TRUE))
+    {
+        ppt->x = (rc.left + rc.right) / 2;
+        ppt->y = (rc.top + rc.bottom) / 2;
+    }
+    ClientToScreen(hWnd, ppt);
+}
+
+static void
+BrFolder_OnContextMenu(BrFolder &info, LPARAM lParam)
+{
+    enum { IDC_EXPAND = 1, ID_FIRSTCMD, ID_LASTCMD = 0xffff };
+    HTREEITEM hSelected = TreeView_GetSelection(info.hwndTreeView);
+    CMINVOKECOMMANDINFOEX ici = { sizeof(ici), CMIC_MASK_PTINVOKE, info.hWnd };
+    ici.fMask |= GetKeyState(VK_SHIFT) < 0 ? CMIC_MASK_SHIFT_DOWN : 0;
+    ici.fMask |= GetKeyState(VK_CONTROL) < 0 ? CMIC_MASK_CONTROL_DOWN : 0;
+    ici.nShow = SW_SHOW;
+    ici.ptInvoke.x = GET_X_LPARAM(lParam);
+    ici.ptInvoke.y = GET_Y_LPARAM(lParam);
+    if ((int)lParam == -1) // Keyboard
+    {
+        GetTreeViewItemContextMenuPos(info.hwndTreeView, hSelected, &ici.ptInvoke);
+    }
+    else // Get correct item for right-click on not current item
+    {
+        TVHITTESTINFO hti = { ici.ptInvoke };
+        ScreenToClient(info.hwndTreeView, &hti.pt);
+        hSelected = TreeView_HitTest(info.hwndTreeView, &hti);
+    }
+    BrItemData *item = BrFolder_GetItemData(&info, hSelected);
+    if (!item)
+        return ; // Not on an item
+    TV_ITEM tvi;
+    tvi.mask = TVIF_STATE | TVIF_CHILDREN;
+    tvi.stateMask = TVIS_EXPANDED;
+    tvi.hItem = hSelected;
+    TreeView_GetItem(info.hwndTreeView, &tvi);
+    IContextMenu *pcm;
+    LPITEMIDLIST child = item->pidlChild; // ATL is stupid and will assert if we take the address of this
+    HRESULT hr = item->lpsfParent->GetUIObjectOf(info.hWnd, 1, &child,
+                                                 IID_IContextMenu, NULL, (void**)&pcm);
+    if (FAILED(hr))
+        return ;
+    HMENU hMenu = CreatePopupMenu();
+    info.pContextMenu = pcm;
+    UINT cmf = ((ici.fMask & CMIC_MASK_SHIFT_DOWN) ? CMF_EXTENDEDVERBS : 0) | CMF_CANRENAME;
+    hr = pcm->QueryContextMenu(hMenu, 0, ID_FIRSTCMD, ID_LASTCMD, CMF_NODEFAULT | cmf);
+    if (hr > 0)
+        _InsertMenuItemW(hMenu, 0, TRUE, 0, MFT_SEPARATOR, NULL, 0);
+    _InsertMenuItemW(hMenu, 0, TRUE, IDC_EXPAND, MFT_STRING,
+        MAKEINTRESOURCEW((tvi.state & TVIS_EXPANDED) ? IDS_COLLAPSE : IDS_EXPAND),
+        MFS_DEFAULT | (tvi.cChildren ? 0 : MFS_GRAYED));
+
+    UINT cmd = TrackPopupMenuEx(hMenu, TPM_RETURNCMD, ici.ptInvoke.x, ici.ptInvoke.y, info.hWnd, NULL);
+    ici.lpVerb = MAKEINTRESOURCEA(cmd - ID_FIRSTCMD);
+    if (cmd != 0 && GetDfmCmd(pcm, ici.lpVerb) == DFM_CMD_RENAME)
+    {
+        PostMessage(info.hwndTreeView, TVM_SELECTITEM, TVGN_CARET, (LPARAM)hSelected);
```

Comment:
Use `TreeView_...` macro.

Reply:
Are those macros ok to use also when someone wants to do **Post**Message() instead of SendMessage?
(PostMessage is superioir when it comes to sending between 2 threads, not sure, whether that is the case here now or in the future)

Reply:
Posting a message will be processed in the future when the message reached.
The instance of the tree item handle or pointer may change in the future.

Reply:
It needs to be Post, if Send had worked I would have used Tree...

Reply:
Why does it need to be Post here?

Reply:
> Why does it need to be Post here?

For whatever reason, it will not expand the item if you Send.

Reply:
I see. Thank you for explanation.

Reply:
Please add a comment for the reason.